internal/framework/provider/functions/encode.go

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package functions
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+
+	"sigs.k8s.io/yaml"
+)
+
+func encodeValue(v attr.Value) (any, error) {
+	switch vv := v.(type) {
+	case basetypes.StringValue:
+		return vv.ValueString(), nil
+	case basetypes.NumberValue:
+		f, _ := vv.ValueBigFloat().Float64()
+		return f, nil
+	case basetypes.Float64Value:
+		return vv.ValueFloat64(), nil
+	case basetypes.Int64Value:
+		return vv.ValueInt64(), nil
+	case basetypes.BoolValue:
+		return vv.ValueBool(), nil
+	case basetypes.ObjectValue:
+		return encodeObject(vv)
+	case basetypes.TupleValue:
+		return encodeTuple(vv)
+	case basetypes.MapValue:
+		return encodeMap(vv)
+	case basetypes.ListValue:
+		return encodeList(vv)
+	case basetypes.SetValue:
+		return encodeSet(vv)
+	default:
+		return nil, fmt.Errorf("tried to encode unsupported type: %T: %v", v, vv)
+	}
+}
+
+func encodeSet(sv basetypes.SetValue) ([]any, error) {
+	elems := sv.Elements()
+	size := len(elems)
+	l := make([]any, size)
+	for i := 0; i < size; i++ {
+		var err error
+		l[i], err = encodeValue(elems[i])
+		if err != nil {
+			return nil, err
+		}
+	}
+	return l, nil
+}
+
+func encodeList(lv basetypes.ListValue) ([]any, error) {
+	elems := lv.Elements()
+	size := len(elems)
+	l := make([]any, size)
+	for i := 0; i < size; i++ {
+		var err error
+		l[i], err = encodeValue(elems[i])
+		if err != nil {
+			return nil, err
+		}
+	}
+	return l, nil
+}
+
+func encodeTuple(tv basetypes.TupleValue) ([]any, error) {
+	elems := tv.Elements()
+	size := len(elems)
+	l := make([]any, size)
+	for i := 0; i < size; i++ {
+		var err error
+		l[i], err = encodeValue(elems[i])
+		if err != nil {
+			return nil, err
+		}
+	}
+	return l, nil
+}
+
+func encodeObject(ov basetypes.ObjectValue) (map[string]any, error) {
+	attrs := ov.Attributes()
+	m := make(map[string]any, len(attrs))
+	for k, v := range attrs {
+		var err error
+		m[k], err = encodeValue(v)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return m, nil
+}
+
+func encodeMap(mv basetypes.MapValue) (map[string]any, error) {
+	elems := mv.Elements()
+	m := make(map[string]any, len(elems))
+	for k, v := range elems {
+		var err error
+		m[k], err = encodeValue(v)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return m, nil
+}
```

Comment:
Another FYI for possible use-cases, the `AllowNullValues` (implicitly set to false for this function) won't prevent null element values (for list, map, set, tuple) or null attribute values (for objects). If that changes how you want to decode things, might be worth adding some test scenarios for that, otherwise you'll get default values from the `Value*()` functions (string for example, null -> "")

Reply:
This is good to know, in our case we want to actually render a `null` in the YAML and not have it result in a default/empty value for that type.

Reply:
Gotcha, so that sounds like you'll want to handle that then.

There's also the possibility that Terraform will know the underlying type (like a string) but not know the underlying value (null or unknown). For null list elements I'd imagine this would be the most common case. I wrote up some initial docs on that here: https://developer.hashicorp.com/terraform/plugin/framework/handling-data/dynamic-data#handling-underlying-null-and-unknown-values

Will probably need to expand on that more in future docs, but let me know if that doesn't make sense 😆